llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,245 @@
+                         "Minimum Iteration Scalar Evolution"
+                      << *MinIterSCEV << "\n");
+    return;
+  }
+  Type *CmpType = LoopCmpInstr.getOperand(0)->getType();
+  Value *LoopAssertValue =
+      Expander.expandCodeFor(MinIterSCEV, CmpType, HeaderTerminator);
+  LLVM_DEBUG(dbgs() << "Loop Iteration Count        : " << MinIterCount << "\n"
+                    << "MinIterValue                : " << *LoopAssertValue
+                    << "\n");
+
+  Value *LHS = BunldedSCEV.IVOperandIndex == 0 ? LoopAssertValue
+                                               : LoopCmpInstr.getOperand(0);
+  Value *RHS = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
+                                               : LoopAssertValue;
+  Value *Cmp = Builder.CreateICmp(LoopCmpInstr.getPredicate(), LHS, RHS);
+
+  // Insert Assumption
+  CallInst *Assumption = Builder.CreateAssumption(Cmp);
+  AC.registerAssumption(dyn_cast<AssumeInst>(Assumption));
+  LLVM_DEBUG(dbgs() << "With Comparator             :" << *Cmp << "\n"
+                    << "Assume                      :" << *Assumption << "\n");
+}
+
+/// Get Minimum Iteration Count of the CurrentLoop if it exists and is greater
+/// than 0
+std::optional<int64_t> getMinIterCount(Loop &CurrentLoop) {
+  BasicBlock *LoopHeader = CurrentLoop.getHeader();
+
+  // Dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  } dbgs() << "LoopIterCountAssumption-Info: Function = "
+           << getFunctionAndBlockNames(*LoopHeader) << "\n");
+
+  std::optional<int64_t> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+  if (!RawMinIterationCount) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions: Loop Iteration "
+                         "Count not provided for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  const int64_t MinIterCount = *RawMinIterationCount;
+  if (MinIterCount <= 0) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Loop Iteration "
+                         "Count is smaller or equal to zero for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << getFunctionAndBlockNames(*LoopHeader) << " ("
+                    << MinIterCount << ")\n");
+
+  if (!CurrentLoop.isRotatedForm())
+    return RawMinIterationCount;
+
+  const int BackedgeCount = MinIterCount - 1;
+  if (BackedgeCount <= 0) {
+    LLVM_DEBUG(
+        dbgs()
+        << "LoopIterCountAssumptions-Warning: Loop is already "
+           "rotated and the Backedge Count is smaller or equal to zero for "
+        << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  return std::optional(BackedgeCount);
```

Comment:
If the loop is already rotated, is there any benefit for us? I mean, we can influence changes in the CFG as you mentioned before (fusing BBs). In this way, I am curious about practical effects (QoR).

I am just pointing this because loop rotation is an premise in several parts of this PR. Could we have a command line option to control if we would like to touch rotated loops and evaluate practical results?

Reply:
we do not have any `do-while` loops  that have `loop iteration count pragmas` attached. Therefore, we do not have any regressions.

Reply:
I am just not sure if we need this, because if or goal is "to help loop rotation" we are inserting unnecessary information. Even if we have such type of loops, `assume` will add a known information.